### PR TITLE
python3+gtk3 in diamond+spud

### PR DESCRIPTION
**Pull request by cianwilson**
Opening a pull request to start a discussion about python3 and gtk3 in spud/diamond.

This branch builds on @jrper's work python3ifying the backend of spud (in the `python3_support` branch) by also updating diamond.  Unfortunately, as there doesn't appear to be any python3 support for pygtk (gtk2) this mostly involves updating gtk2 to gtk3, for which there is support in python3 and backwards compatible support to python2.  Hence this pull request ends up mostly being about switching to gtk3, with the side effect of making diamond work with python3.

I've tested this as much as I can but there are a few features that I know are untested and therefore definitely broken:
* plugins - are these used any more? does anyone have an example I can test?
* validate schematron - not sure what this does, any example out there?
* validate xml - this is currently broken in the same way as the master branch in the tests I've done
* diff - works some of the time (and I've fixed it in those circumstances) but when it doesn't it's broken in the same way as on master
* find - though part of the UI description this was disabled within diamond so I haven't reimplemented it at all

One minor note: in the upgrade we've lost the default zebra striping of the options tree.  I didn't reimplement this but it appears possible using a css file.

I'd appreciate it if others could test and report back what's broken.

To install the new dependencies...

## Ubuntu (pure apt)

I tested this on a clean ubuntu 16.04 docker image but I'm assuming here that you're starting from a computer that already builds spud (so I'm not giving instructions for installing gcc, git etc.).

Install the following:
```
sudo apt install -y python-gi \
                               python-future \
                               python-configparser \
                               gir1.2-gtksource-3.0 \
                               gir1.2-gtk-3.0
```
for python2, or:
```
sudo apt install -y python3-gi \
                               python3-future \
                               gir1.2-gtksource-3.0 \
                               gir1.2-gtk-3.0
```
for python3.

## Ubuntu (using pip, e.g. within a virtualenv or conda env or just using pip or conda)

Based partially on these instructions: http://pygobject.readthedocs.io/en/latest/devguide/dev_environ.html#linux-pipsi
and tested on a clean ubuntu 16.04 docker image (again, not giving instructions for installing git etc.).

I haven't quite worked out the conda packages for this yet so below I'm just using pip but I'll update if I have any more luck.  In particular I can't find a `gtksource` conda package.  Additionally, if your distribution of conda has cairo installed already (e.g. anaconda by default ships with it, use `conda list` to check) you'll need to remove that, if possible, using `conda remove cairo` as it is not compatible with pip's `pycairo` (installed automatically via `pygobject` below).

```
sudo apt install -y libgirepository1.0-dev build-essential \
                               libbz2-dev libreadline-dev libssl-dev zlib1g-dev libsqlite3-dev wget \
                               curl llvm libncurses5-dev libncursesw5-dev xz-utils tk-dev libcairo2-dev \
                               gir1.2-gtksource-3.0 gir1.2-gtk-3.0
pip install pygobject lxml future
```

## Mac OSX

I tested this on a high sierra machine but unfortunately I don't have access to a clean one so can't tell if this is a complete description.  I'm assuming use of the brew package manager.

I think:
```
brew install pygobject3 adwaita-icon-theme gtksourceview3
```
will generally be sufficient, or:
```
brew install pygobject3 --with-python@2
brew install adwaita-icon-theme gtksourceview3
```
if you want python2 support.

Additionally, if you don't have them already:
```
pip install configparser future
```
where you should switch to pip[2/3] if necessary for your python distribution.

### CODE REVIEW

**Comment by cianwilson** [comment]
Thanks for the bionic update @Patol75.  I'm back at my desktop now, which is running 18.04 so I'll try it out here too.  On 16.04 I think the pycairo dependency is built by pip when building pygobject (at least that's what the output of the command looks like to me - `pip install pygobject` looks like it's failing but actually isn't - it seems to try once, fail to find a cairo module, goes away and builds pycairo, returns to pygobject and succeeds the second time around).  It is an option to `pip install pycairo` as well (on 16.04 at least) if you're looking to keep this contained in a virtual env and not use apt.

I don't know how the branching of spud works from within the fluidity directory.  I'd recommend cloning and installing this branch of the spud repo directly.

Thanks for the CSS info.  It wasn't that I didn't know how to do it, it was just that I didn't think it was important enough to fix and wanted to open it up to opinions.  I think we only had it before because it was on by default.  I personally quite like the look without it but I'm open to other opinions.

**Comment by cianwilson** [comment]
Confirmed that for me `pip install pygobject` seems sufficient on ubuntu 18.04 (bionic) inside a python3 virtual env without `sudo apt install python3-cairo-dev`.  The tail of my `pip install pygobject lxml future` command looks like:
```
  ModuleNotFoundError: No module named 'cairo'
  
  ----------------------------------------
  Failed building wheel for pygobject
  Running setup.py clean for pygobject
  Running setup.py bdist_wheel for future ... done
  Stored in directory: /home/cwilson/.cache/pip/wheels/bf/c9/a3/c538d90ef17cf7823fa51fc701a7a7a910a80f6a405bf15b1a
  Running setup.py bdist_wheel for pycairo ... done
  Stored in directory: /home/cwilson/.cache/pip/wheels/96/d7/27/11325e387f45d69372212d0da63c366fe0db450c2cab978ad2
Successfully built future pycairo
Failed to build pygobject
Installing collected packages: pycairo, pygobject, lxml, future
  Running setup.py install for pygobject ... done
Successfully installed future-0.16.0 lxml-4.2.4 pycairo-1.17.1 pygobject-3.28.3
```

**Comment by stephankramer** [comment]
Thanks for the heads up! Was just about to say, everything appears to work fine for me (Debian buster, python2 and python3)...

**Comment by cianwilson** [comment]
Thanks @stephankramer!

Save failure should be fixed now.

Also reimplemented Ctrl+s etc. for save and fixed the lost/added xml window.

Really hard to catch everything when it depends on clicking on it before seeing it fail!

**Comment by Patol75** [comment]
I have one small bug (Ubuntu 18.04.1 LTS), when I am in an algorithm (tensor_python_diagnostic) node, changing the material_phase_support attribute from single to multiple and vice-versa does not always work, would it be using the mouse or the keys. Actually, I think it works only the first time. If I want to change it again, I need to click on another node and then come back to the algorithm (tensor_python_diagnostic) node.
Apart from this, tremendous work ! :)

**Comment by cianwilson** [comment]
Thanks @Patol75, when this happens do you get any output to the terminal from which you're running diamond?

**Comment by cianwilson** [comment]
@Patol75:  confirmed your bug but it was unrelated to this pull request and occurred in old versions of diamond as well.  By the time I'd figured that out though I knew how to fix it so pushed a fix built upon this branch to the `gtk3_python3_getattr` branch.  Could you please check it out and confirm it fixes your issue?  You can continue testing all features of this PR in that branch if you want. Thanks!

**Comment by Patol75** [comment]
Confirmed bug fixed on the branch ```gtk3_python3_getattr```. Thanks for that @cianwilson. No other issue on my side. :)

**Comment by drhodrid** [comment]
Hi all - where do things stand with this? Can we get it merged?

**Comment by angus-g** [comment]
Just chiming in that I can build the python extension and diamond with 2.7 and 3.6 on this. The changes themselves all look pretty sensible too.

**Comment by drhodrid** [comment]
@tmbgreaves - Now that this has been approved, what has to be done to get it merged? Any chance we could work towards a merge in the near future?

**Comment by tmbgreaves** [comment]
Hi @drhodrid, sorry for missing this one. Now caught it and will do a rebase and make sure Jenkins is happy.

**Comment by tmbgreaves** [comment]
OK -- that's Jenkins fixed, a new base image pushed with just the packages needed for spud, and testing running green. Since this has required some additions to the branch, I would be grateful if someone other than me could take a quick look and make sure I've not messed up anything critical. Once someone else has cast an eye over it, if they're happy, I'm more than happy for the merge button to be pressed.

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    stages {
+        stage('Configuring') {   
+            steps { 
+                sh './configure --prefix=/home/spud' 
+            }
+        }    
+        stage('Building for python2') {       
+            steps { 
+                sh 'make -j'
+                sh 'make doc'
```
Comment (angus-g) [inline Jenkinsfile]: Leave for the "building documentation" stage?

Comment (tmbgreaves) [reply Jenkinsfile]: Good spot - thank you!

**Comment by angus-g** [comment]
Diamond looks fine, and seems to work here (doesn't look like it got changed since my last review anyway). However, I am seeing the `python_diagnostic_field` test fail in fluidity (Python 3.7.3):

```
Traceback (most recent call last):
  File "<string>", line 10, in <module>
SpudKey.error: Error: The specified option is not present in the dictionary in get option
 Python error, Python string was:
 X=state.vector_fields["Coordinate"]

for i in range(field.node_count):
    X_n=X.node_val(i)
    field.set(i, X_n[0]*X_n[1])

import sys
if sys.version_info[1] >= 7:
    import libspud
    name = libspud.get_option("/simulation_name")
    assert name == "python"
*** ERROR ***
Error message: Dying
```

**Comment by jrper** [comment]
@angus-g That test is extremely flakey and can fail in multiple ways, but have you checked that your system doesn't have a shared build of libspud on it, or that if it does, neither fluidity nor the python libspud module you're importing happen to link against it? That's probably the easiest way to end up with multiple instances of the libspud manager which then fail to interact.

**Comment by angus-g** [comment]
@jrper I've checked that, and I've only got `libspud.a`, `libspud.so` and `libspud.so.0` in fluidity's lib directory, and nowhere else. Looks like perhaps something changed around Python modules and shared libraries:

Fluidity links against the static `libspud.a`, but the Python module is `python/libspud.cpython-37m-x86_64-linux-gnu.so` for me, which links against the spud shared library:
```
$ ldd python/libspud.cpython-37m-x86_64-linux-gnu.so
...
        libspud.so.0 => not found
```

If I link fliudity against the libspud shared library (changing `LIBS` in the Makefile to use `-lspud` rather than `libspud.a`), the test passes for me. Unsurprisingly, doing things the other way around, by linking against libspud.a to build the Python module, doesn't work.

Compare this to the behaviour under Python 2.7.
```
$ ldd python/libspud.so
...
        libspud.so.0 => /home/angusgibson/fluidity/libspud/libspud.so.0 (0x00007f1a522d5000)
```

It may be a red herring, but it's interesting that the older version seems to have the `RUNPATH` set (not that it matters because `LD_LIBRARY_PATH` gets set anyway).

**Pull request merged**